src/Language/PureScript/Linter.hs

```diff
@@ -129,3 +132,136 @@ lint (Module _ _ mn ds _) = censor (addHint (ErrorInModule mn)) $ mapM_ lintDecl
     | name `S.member` s = (s, errorMessage' ss (mkError name))
     | otherwise = (S.insert name s, mempty)
 
+
+
+lintUnused :: forall m. (MonadWriter MultipleErrors m) => Module -> m ()
+lintUnused (Module modSS _ mn modDecls exports) =
+  censor (addHint (ErrorInModule mn)) $ do
+    topVars <- traverse lintDeclaration modDecls
+    let allVars = S.unions topVars
+    case exports of
+      Nothing ->
+        pure ()
+      Just exports' -> do
+        let n (ValueRef _ ident) = Just ident
+            n _ = Nothing
+            exportIds = S.fromList $ mapMaybe n exports'
+            expectedUsedDecls = S.fromList (mapMaybe getDeclIdent modDecls) `S.difference` exportIds
+            newErrs = mconcat $ map (errorMessage' modSS . UnusedDeclaration) $ S.toList $ expectedUsedDecls `S.difference` allVars
+        tell newErrs
+        pure ()
+  where
+  getDeclIdent :: Declaration -> Maybe Ident
+  getDeclIdent (ValueDeclaration vd) = Just (valdeclIdent vd)
+  getDeclIdent (ExternDeclaration _ ident _) = Just ident
+  getDeclIdent (TypeInstanceDeclaration _ _ _ ident _ _ _ _) = Just ident
+  getDeclIdent BindingGroupDeclaration{} = internalError "lint: binding groups should not be desugared yet."
+  getDeclIdent _ = Nothing
+
+  lintDeclaration :: Declaration -> m (S.Set Ident)
+  lintDeclaration declToLint = do
+    let (vars, errs) = goDecl declToLint
+    tell errs
+    pure vars
+    where
+
+    goDecl :: Declaration -> (S.Set Ident, MultipleErrors)
+    goDecl d@(ValueDeclaration vd) =
+        let allExprs = concatMap unguard $ valdeclExpression vd
+            bindNewNames = S.fromList (concatMap binderNames $ valdeclBinders vd)
+            ss = declSourceSpan d
+            (vars, errs) = removeAndWarn ss bindNewNames $ foldr1 combine $ map (go ss) allExprs
+            errs' = addHint (ErrorInValueDeclaration $ valdeclIdent vd) errs
+        in
+          (vars, errs')
+
+    goDecl _ = mempty
+
+    go :: SourceSpan -> Expr -> (S.Set Ident, MultipleErrors)
```

Comment:
If `everythingWithContextOnValues` let you get a hold of `PositionedValue`, could this be implemented in terms of that traversal? Much of this traversal looks like it's reimplementing that, but the only difference is the context is a source span, which you can't get with the current implementation. I don't think it would break anything to expose that.

Reply:
No, the main issue with `everythingWithContextOnValues` is that at the point of binders, we don't want to do something on the binder and the expression under its scope separately and combine them, we want to take the result on the scoped expression then look at that together with the binder. I guess it would likely be possible to bake more into the context and do all the logic in the combine, but that seems rather contorted, really this is another traversal for capture-aware operations.

(Aside, I originally followed the logic for the type-level version of the same, which inverts to go top-down with the set of so far unused variables - now that was a bunch of make-work...)